src/lib/pool-info.ts

```diff
@@ -358,6 +377,85 @@ export class PoolInfo {
     await this.setLatestBlockNumber();
   }
 
+  // This is not fully correct as the deposits done on the same block after the reward tx is
+  // also considered. epochEndBlock is used instead of epochEndBlock - 1 as with epochEndBlock
+  // the earnedPSP is a lower bound while on the other case its an upper bound.
+  // TODO: memorize this on db to avoid archive calls
+  async fetchEarnedPSPEpoch(userAddress: string, epoch: number): Promise<{[poolAddress: string]: bigint}> {
+    const epochInfo = EpochDetails[this.network][epoch];
+    if(!epochInfo)
+      throw new Error(`Epoch details not found.`);
+
+    const epochEndBlockNumber = epochInfo.endBlockNumber;
+
+    const pools = this.poolConfigs.filter(p => p.poolReleaseBlockNumber <= epochEndBlockNumber);
+
+    if(!this.poolSPSPSupply[epochEndBlockNumber]) {
+      const multiCallData = pools.map(p => {
+        return {
+          target: p.address,
+          callData: this.spspInterface.encodeFunctionData('totalSupply'),
+        };
+      });
+      const rawResult = await this.multicallContract.functions.aggregate(
+        multiCallData,
+        { blockTag: epochEndBlockNumber },
+      );
+      const SPSPSupply: {[address: string]: bigint} = {};
+      pools.forEach((p, i) => SPSPSupply[p.address.toLowerCase()] = BigInt(
+        this.spspInterface
+          .decodeFunctionResult('totalSupply', rawResult.returnData[i])
+          .toString(),
+      ));
+      this.poolSPSPSupply[epochEndBlockNumber] = SPSPSupply;
+    }
+    const SPSPSupply = this.poolSPSPSupply[epochEndBlockNumber];
+
+    const multiCallData = pools.map(p => {
+      return {
+        target: p.address,
+        callData: this.erc20Interface.encodeFunctionData('balanceOf', [
+          userAddress,
+        ]),
+      };
+    });
+
+    const rawResult = await this.multicallContract.functions.aggregate(
```

Comment:
we can memoize this too, on memory for now

Reply:
An attacker can simply call with random address which will bombard the memory

Reply:
Very fair point